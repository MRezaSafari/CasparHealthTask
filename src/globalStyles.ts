import { createGlobalStyle } from "styled-components";

const GlobalStyle = createGlobalStyle`
  body {
    margin: 0;
    padding: 30px 0;
    background: #f1f1f1;
    color: #000;
    font-family: 'Open Sans', sans-serif;
  }

  *{
    box-sizing: border-box;
  }

  .container {
    max-width: 1280px;
    margin: auto;
    position: relative;
    padding: 0 15px;
  }

  ul{
    list-style: none;
    padding: 0;
  }

  a{
    text-decoration: none;
  }

 h1,h2,h3,h4,h5,h6{
    margin: 0;
  }
`;

export default GlobalStyle;
