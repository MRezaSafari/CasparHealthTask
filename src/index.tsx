import * as React from "react";
import { createRoot } from "react-dom/client";

import AppRoot from "./app";
import { BrowserRouter } from "react-router-dom";
import ErrorBoundary from "./components/error-boundary";
import { HelmetProvider } from "react-helmet-async";

let rootElement: unknown = null;

if (!rootElement) {
  rootElement = document.getElementById("root") as HTMLElement;
  const root = createRoot(rootElement as HTMLDivElement);
  root.render(
    <React.StrictMode>
      <ErrorBoundary>
        <HelmetProvider>
          <BrowserRouter>
            <AppRoot />
          </BrowserRouter>
        </HelmetProvider>
      </ErrorBoundary>
    </React.StrictMode>
  );
}

// @ts-ignore
if (module.hot) {
  // @ts-ignore
  module.hot.accept();
}
