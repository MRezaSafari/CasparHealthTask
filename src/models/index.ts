export * from "./patients.model";
export * from "./table.model";
export * from "./modal.model";
