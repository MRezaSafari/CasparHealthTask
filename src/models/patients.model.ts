export type Gender = "Male" | "Female";

export interface IPatient {
  patient_id: number;
  first_name: string;
  last_name: string;
  email: string;
  gender: Gender;
  age: number;
  avatar: string;
}

export interface IPatientsStore {
  patients: IPatient[];
  setPatients: (patients: IPatient[]) => void;
  filters: IFilters;
  setFilters: (filters: IFilters) => void;
  reload: () => void;
}

export interface IFilters {
  isInitial: boolean;
  query: string;
  age: string;

  gender: Gender | 'All';
}
