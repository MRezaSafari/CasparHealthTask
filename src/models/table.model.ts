import { ReactNode } from "react";

type ColumnType = "string" | "boolean";
type OrderDirection = "ASC" | "DESC";

interface IColumnTemplate<T> {
  title: string;
  width: string;
  type?: ColumnType;
  valueKey?: string;
  sortable: boolean;
  render?: (row: T) => ReactNode;
}

interface ITableProps<T> {
  columns: IColumnTemplate<T>[];
  data: T[] | undefined;
  onSort?: (column: keyof T, direction: OrderDirection) => void;
  onRowClick?: (row: T) => void;
}

export type { IColumnTemplate, ColumnType, ITableProps, OrderDirection };
