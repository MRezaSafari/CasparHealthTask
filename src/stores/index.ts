export { default as usePatientsStore } from "./patients/patients.store";
