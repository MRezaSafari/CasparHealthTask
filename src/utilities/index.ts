export { default as debounce } from './debounce/debounce';
export * from './patients/patients.utils';