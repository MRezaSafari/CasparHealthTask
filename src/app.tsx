import { Route, Routes } from "react-router-dom";
import Home from "./pages/home";
import { Helmet } from "react-helmet-async";
import GlobalStyle from "./globalStyles";
import { usePatientsStore } from "./stores";
import { useEffect } from "react";

const AppRoot = () => {
  const { initiate } = usePatientsStore();

  useEffect(() => {
    initiate();
  }, []);
  
  return (
    <>
      <Helmet>
        <meta charSet="utf-8" />
        <title>Caspar Health Interview Task</title>
      </Helmet>
      <main>
        <GlobalStyle />
        <div className="container">
          <Routes>
            <Route index element={<Home />} />
          </Routes>
        </div>
      </main>
    </>
  );
};

export default AppRoot;
