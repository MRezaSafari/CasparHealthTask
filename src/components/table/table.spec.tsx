import { render } from "@testing-library/react";
import Table from "./index";
import { IColumnTemplate, IPatient } from "src/models";

const columns: IColumnTemplate<IPatient>[] = [
  {
    sortable: true,
    title: "First name",
    width: "20%",
    valueKey: "first_name",
    type: "string",
  },
  {
    sortable: true,
    title: "Last name",
    width: "20%",
    valueKey: "last_name",
    type: "string",
  },
];

const mockData: IPatient[] = [
  {
    patient_id: 260,
    first_name: "Berke",
    last_name: "Tett",
    email: "btett77@vimeo.com",
    gender: "Male",
    age: 67,
    avatar: "http://dummyimage.com/221x100.png/ff4444/ffffff",
  },
  {
    patient_id: 261,
    first_name: "Catlaina",
    last_name: "Estrella",
    email: "cestrella78@google.it",
    gender: "Female",
    age: 40,
    avatar: "http://dummyimage.com/113x100.png/ff4444/ffffff",
  },
];

describe("Table", () => {

  it("should render successfully", () => {
    const {container} = render(<div><Table columns={columns} data={[]} /></div>);
    expect(container.getElementsByTagName("table")[0]).toBeInTheDocument();
  });

  it("should match the snapshot", () => {
    const { container } = render(<Table columns={columns} data={[]} />);
    expect(container).toMatchSnapshot();
  });

  it("should match the snapshot with data", () => {
    const { container } = render(<Table columns={columns} data={mockData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render the correct number of rows", () => {
    const { container } = render(<Table columns={columns} data={mockData} />);
    expect(container.querySelectorAll("tbody > tr").length).toBe(2);
  });
});
