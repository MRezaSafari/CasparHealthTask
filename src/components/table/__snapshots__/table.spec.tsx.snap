// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should match the snapshot 1`] = `
<div>
  <section>
    <table
      cellpadding="0"
      cellspacing="0"
      class="sc-bdnyFh eSGjeE"
      data-testid="table"
    >
      <thead>
        <tr>
          <td
            width="20%"
          >
            <div
              class="sc-gtsqUy cbCKTL"
            >
              <span>
                First name
              </span>
              <ul>
                <li>
                  <svg
                    class="sort-asc"
                    fill="none"
                    height="15"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="15"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6 15l6 -6l6 6"
                    />
                  </svg>
                </li>
                <li>
                  <svg
                    class="tabler-icon tabler-icon-chevron-down"
                    data-testid="sort-desc"
                    fill="none"
                    height="15"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="15"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6 9l6 6l6 -6"
                    />
                  </svg>
                </li>
              </ul>
            </div>
          </td>
          <td
            width="20%"
          >
            <div
              class="sc-gtsqUy cbCKTL"
            >
              <span>
                Last name
              </span>
              <ul>
                <li>
                  <svg
                    class="sort-asc"
                    fill="none"
                    height="15"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="15"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6 15l6 -6l6 6"
                    />
                  </svg>
                </li>
                <li>
                  <svg
                    class="tabler-icon tabler-icon-chevron-down"
                    data-testid="sort-desc"
                    fill="none"
                    height="15"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="15"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6 9l6 6l6 -6"
                    />
                  </svg>
                </li>
              </ul>
            </div>
          </td>
        </tr>
      </thead>
    </table>
    <div
      class="sc-dlniIP eYvRRJ"
    >
      No Data!
    </div>
  </section>
</div>
`;

exports[`Table should match the snapshot with data 1`] = `
<div>
  <section>
    <table
      cellpadding="0"
      cellspacing="0"
      class="sc-bdnyFh eSGjeE"
      data-testid="table"
    >
      <thead>
        <tr>
          <td
            width="20%"
          >
            <div
              class="sc-gtsqUy cbCKTL"
            >
              <span>
                First name
              </span>
              <ul>
                <li>
                  <svg
                    class="sort-asc"
                    fill="none"
                    height="15"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="15"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6 15l6 -6l6 6"
                    />
                  </svg>
                </li>
                <li>
                  <svg
                    class="tabler-icon tabler-icon-chevron-down"
                    data-testid="sort-desc"
                    fill="none"
                    height="15"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="15"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6 9l6 6l6 -6"
                    />
                  </svg>
                </li>
              </ul>
            </div>
          </td>
          <td
            width="20%"
          >
            <div
              class="sc-gtsqUy cbCKTL"
            >
              <span>
                Last name
              </span>
              <ul>
                <li>
                  <svg
                    class="sort-asc"
                    fill="none"
                    height="15"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="15"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6 15l6 -6l6 6"
                    />
                  </svg>
                </li>
                <li>
                  <svg
                    class="tabler-icon tabler-icon-chevron-down"
                    data-testid="sort-desc"
                    fill="none"
                    height="15"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="15"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6 9l6 6l6 -6"
                    />
                  </svg>
                </li>
              </ul>
            </div>
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            width="20%"
          >
            Berke
          </td>
          <td
            width="20%"
          >
            Tett
          </td>
        </tr>
        <tr>
          <td
            width="20%"
          >
            Catlaina
          </td>
          <td
            width="20%"
          >
            Estrella
          </td>
        </tr>
      </tbody>
    </table>
  </section>
</div>
`;
